packages/swap/src/quoting/Quoter.ts

```diff
@@ -0,0 +1,234 @@
+import BigNumber from 'bignumber.js';
+import { randomUUID } from 'crypto';
+import { Subject, Subscription, filter } from 'rxjs';
+import { Server } from 'socket.io';
+import { findPrice, type SwapInput } from '@/amm-addon';
+import { getPoolsNetworkFeeHundredthPips } from '@/shared/consts';
+import { InternalAsset } from '@/shared/enums';
+import { getHundredthPipAmountFromAmount } from '@/shared/functions';
+import { SwapFee } from '@/shared/schemas';
+import { QuotingSocket } from './authenticate';
+import Leg from './Leg';
+import {
+  Leg as MarketMakerLeg,
+  MarketMakerQuote,
+  MarketMakerQuoteRequest,
+  marketMakerResponseSchema,
+} from './schemas';
+import { Pool } from '../client';
+import env from '../config/env';
+import { getAssetPrice } from '../pricing';
+import { buildFee } from '../utils/fees';
+import logger from '../utils/logger';
+import { percentDiff } from '../utils/math';
+import { getSwapRate } from '../utils/statechain';
+
+export type QuoteType = 'pool' | 'market_maker';
+
+type Quote = { marketMaker: string; quote: MarketMakerQuote };
+
+const getPriceFromQuotesAndPool = async (leg: Leg, input: SwapInput) => {
+  const { swappedAmount, remainingAmount } = await findPrice(input);
+
+  if (remainingAmount !== 0n) {
+    // ignoring fees from the pool
+    const { outputAmount } = await getSwapRate({
+      ...leg.toPoolJSON(),
+      amount: remainingAmount,
+    });
+
+    return swappedAmount + outputAmount;
+  }
+
+  return swappedAmount;
+};
+
+export const approximateIntermediateOutput = async (asset: InternalAsset, amount: string) => {
+  const price = await getAssetPrice(asset);
+
+  if (typeof price !== 'number') return null;
+
+  return BigInt(new BigNumber(amount).times(price).shiftedBy(-2).toFixed(0));
```

Comment:
do we need to shift based on the decimals of the `asset` here? i think the return value is an usdc amount (and therefore needs to have 6 decimals), right?

Reply:
good catch, it's the USDC decimals - input asset decimals. I've added tests specifically for this method.